### ISSUE TITLE
Using a param-block inside MockWith behaves strangely

### ISSUE THREAD

**Issue by johlju** [pester/Pester#1554]
## 1. General summary of the issue

Using a `param()` block inside a `-MockWith` behaves strangely.

## 2. Describe Your Environment

Using Pester 5 **RC9**

```powershell
Pester version     : 5.0.0 C:\source\SqlServerDsc\output\RequiredModules\Pester\5.0.0\Pester.psd1
PowerShell version : 7.0.1
OS version         : Microsoft Windows NT 10.0.19041.0
```

## 3. Expected Behavior

Either change the documentation that using `param()`-block should no longer be used (breaking change), otherwise the expected behavior is that this should work (if it still has value?).

## 4.Current Behavior

The following mockup:

```powershell
BeforeAll {
    function Get-Something
    {
        [CmdletBinding()]
        param
        (
            [Parameter()]
            $MyParam1,

            [Parameter()]
            $MyParam2
        )
    }

    function Set-Something
    {
        [CmdletBinding()]
        param ( )

        Get-Something -MyParam1 'SomeValue'
    }
}

Describe 'MyTest2' {
    Context 'Fails using one parameter' {
        BeforeAll {
            Mock Get-Something -MockWith {
                param
                (
                    [Parameter()]
                    [System.String]
                    $MyParam1
                )

                return $MyParam1
            }
        }

        It 'MyIt' {
            Set-Something  | Should -Be 'SomeValue'
        }
    }

    Context 'Works adding another dummy parameter' {
        BeforeAll {
            Mock Get-Something -MockWith {
                param
                (
                    [Parameter()]
                    [System.String]
                    $MyParam1,

                    [Parameter()]
                    [System.String]
                    $UnknownParam
                )

                return $MyParam1
            }
        }

        It 'MyIt' {
            Set-Something  | Should -Be 'SomeValue'
        }
    }

    Context 'Works by removing [Parameter()] from first parameter' {
        BeforeAll {
            Mock Get-Something -MockWith {
                param
                (
                    [System.String]
                    $MyParam1
                )

                return $MyParam1
            }
        }

        It 'MyIt' {
            Set-Something  | Should -Be 'SomeValue'
        }
    }
}
```

fails with:

```
Describing MyTest2
 Context Fails using one parameter
   [-] MyIt 11ms (10ms|1ms)
    ParameterBindingException: A positional parameter cannot be found that accepts argument '$null'.

 Context Works adding another dummy parameter
   [+] MyIt 5ms (4ms|1ms)
 Context Works by removing [Parameter()] from first parameter
   [+] MyIt 5ms (5ms|1ms)
Tests completed in 392ms
Tests Passed: 2, Failed: 1, Skipped: 0 NotRun: 0
```

## 5. Possible Solution

See expected behavior)

## 6. Context

This is not an issue for me as I never had to use a parameter-block inside a `MockWith` scriptblock. This was an very old test that was using this. By removing the entire parameter block from the `MockWith` scriptblock the test worked which was enough fix for me. :)

**Comment by fflaten** [comment]
Looks like it's caused by a hardcoded `$null` value for ArgumentList in the Mock-template (line 143). It is later on cast to an `[object[]]` which results in an array with a single null-valued item.

https://github.com/pester/Pester/blob/8d555269cfcd8e3aee13d1efb10f962afc922f96/src/functions/Mock.ps1#L141-L144

```
PS > $a = $null
PS > [object[]]$o = @($a)
PS > $o.Count
1
```

When this is later on in the following line passed to the Mock for execution, the parameter binding occurs:
https://github.com/pester/Pester/blob/8d555269cfcd8e3aee13d1efb10f962afc922f96/src/functions/Mock.ps1#L950

Seems to work by a simple change of L143 in the Mock template to `$MyInvocation.MyCommand.Mock.Args = @()` as we always expect `[object[]]` everywhere it's used.

The thing I'm stuck on is why it causes ParameterBindingException instead of generating a `$args = @($null)` inside the Mock which no-one would notice.

**Comment by fflaten** [comment]
> The thing I'm stuck on is why it causes ParameterBindingException instead of generating a `$args = @($null)` inside the Mock which no-one would notice.

Found it. `$args` are not allowed in advanced functions. As soon as `[CmdletBinding()]` or `[Parameter()]` is used, you need to use the `[Parameter(ValueFromRemainingArguments=$true)]` option on a parameter to silently catch remaining arguments. Thought that was optional to rename the variable. Learn something new every day 🙂

**Comment by fflaten** [comment]
I've submitted a PR to fix the issue based on the findings above. However, while I was writing different tests scenarios which were not working the way I originally expected, I began to wonder if this is a even a supported scenario.

Should you be able to use declare parameters in a Mock at all? Isn't the point of a mock to change the execution logic while accepting the same parameters as the original function/cmdlet? @nohwnd ?
